transports/quic/src/transport.rs

```diff
@@ -181,28 +214,59 @@ impl<P: Provider> Transport for GenTransport<P> {
                 };
                 &mut dialer.state
             }
-            1 => &mut listeners[0].dialer_state,
-            _ => {
-                // Pick any listener to use for dialing.
-                // We hash the socket address to achieve determinism.
-                let mut hasher = DefaultHasher::new();
-                socket_addr.hash(&mut hasher);
-                let index = hasher.finish() as usize % listeners.len();
-                &mut listeners[index].dialer_state
-            }
+            Some(listener) => &mut listener.dialer_state,
         };
-        Ok(dialer_state.new_dial(socket_addr, self.handshake_timeout, version))
+        Ok(dialer_state.new_dial(socket_addr, handshake_timeout, version))
     }
 
     fn dial_as_listener(
         &mut self,
         addr: Multiaddr,
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
-        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
-        // `addr`. See DCUtR specification below.
-        //
-        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
-        Err(TransportError::MultiaddrNotSupported(addr))
+        let (socket_addr, _version, peer_id) = self.remote_multiaddr_to_socketaddr(addr.clone())?;
+        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr))?;
+
+        let endpoint_channel = match self.eligible_listener(&socket_addr) {
+            None => {
+                return Err(TransportError::Other(
+                    Error::NoActiveListenerForDialAsListener,
+                ));
+            }
+            Some(listener) => listener.endpoint_channel.clone(),
+        };
+
+        let hole_puncher = hole_puncher(endpoint_channel, socket_addr, self.handshake_timeout);
+
+        let (sender, receiver) = oneshot::channel();
+
+        match self.hole_punch_attempts.entry(socket_addr) {
+            Entry::Occupied(mut sender_entry) => {
+                if sender_entry.get().is_canceled() {
+                    sender_entry.insert(sender);
+                } else {
+                    return Err(TransportError::Other(Error::HolePunchInProgress(
+                        socket_addr,
+                    )));
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(sender);
+            }
+        };
+
+        Ok(Box::pin(async move {
+            futures::pin_mut!(hole_puncher);
+            match futures::future::select(receiver, hole_puncher).await {
+                Either::Left((message, _)) => {
+                    let (inbound_peer_id, connection) = message.unwrap().await?;
+                    if inbound_peer_id != peer_id {
+                        log::warn!("inbound connection from ({socket_addr}, {inbound_peer_id} incorrectly matched with outbound hole punch for ({socket_addr}, {peer_id}).");
```

Comment:
Should this be a panic instead of a warning?

Reply:
We don't actually need to check this here, we verify that on the swarm level.

Reply:
Can you point me to where this is checked?

Reply:
See https://github.com/libp2p/rust-libp2p/blob/2910c985e8ab75b52d36701b9c7717ebf0fd8585/swarm/src/connection/pool.rs#L668-L685.